### PR TITLE
MINOR: Replaced Utils.join() with JDK API.

### PR DESCRIPTION
**Pull request by chiacyu**
Since there's [String.join()](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.Iterable-) in  JDK, this pr replaced the current implemented join() with the String.join().

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java [hunk 1]
```diff
@@ -729,7 +729,7 @@ public void testConfigValidationAllOverride() {
         String idempotenceConfigKey = producerOverrideKey(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG);
         config.put(idempotenceConfigKey, "true");
         String bootstrapServersConfigKey = producerOverrideKey(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG);
-        config.put(bootstrapServersConfigKey, "SASL_PLAINTEXT://localhost:12345,SASL_PLAINTEXT://localhost:23456");
+        config.put(bootstrapServersConfigKey, "SASL_PLAINTEXT://localhost:12345, SASL_PLAINTEXT://localhost:23456");
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java]: could you please revert this unrelated change?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 2]
```diff
@@ -1491,7 +1490,7 @@ public void assign(Collection<TopicPartition> partitions) {
             // See the ApplicationEventProcessor.process() method that handles this event for more detail.
             applicationEventHandler.add(new AssignmentChangeEvent(subscriptions.allConsumed(), time.milliseconds()));
 
-            log.info("Assigned to partition(s): {}", join(partitions, ", "));
+            log.info("Assigned to partition(s): {}", String.join(", ", Arrays.toString(partitions.toArray())));
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: `Arrays.toString` will add `[]` so that is a bit different to origin message. Maybe you can use the lambda to generate same message.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResourceTest.java [hunk 3]
```diff
@@ -106,7 +106,7 @@ public class ConnectorPluginsResourceTest {
         PARTIAL_PROPS.put("name", "test");
         PARTIAL_PROPS.put("test.string.config", "testString");
         PARTIAL_PROPS.put("test.int.config", "1");
-        PARTIAL_PROPS.put("test.list.config", "a,b");
+        PARTIAL_PROPS.put("test.list.config", "a, b");
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResourceTest.java]: ditto. please revert unrelated change

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResourceTest.java [hunk 4]
```diff
@@ -188,7 +188,7 @@ public class ConnectorPluginsResourceTest {
         partialConfigs.add(configInfo);
 
         configKeyInfo = new ConfigKeyInfo("test.list.config", "LIST", true, null, "HIGH", "Test configuration for list type.", "Test", 2, "LONG", "test.list.config", Collections.emptyList());
-        configValueInfo = new ConfigValueInfo("test.list.config", "a,b", asList("a", "b", "c"), Collections.emptyList(), true);
+        configValueInfo = new ConfigValueInfo("test.list.config", "a, b", asList("a", "b", "c"), Collections.emptyList(), true);
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResourceTest.java]: ditto. please revert unrelated change

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 5]
```diff
@@ -132,7 +132,7 @@ static Set<ConsumerGroupState> consumerGroupStatesFromString(String input) {
         Set<ConsumerGroupState> parsedStates = Arrays.stream(input.split(",")).map(s -> ConsumerGroupState.parse(s.trim())).collect(Collectors.toSet());
         if (parsedStates.contains(ConsumerGroupState.UNKNOWN)) {
             Collection<ConsumerGroupState> validStates = Arrays.stream(ConsumerGroupState.values()).filter(s -> s != ConsumerGroupState.UNKNOWN).collect(Collectors.toList());
-            throw new IllegalArgumentException("Invalid state list '" + input + "'. Valid states are: " + Utils.join(validStates, ", "));
+            throw new IllegalArgumentException("Invalid state list '" + input + "'. Valid states are: " + String.join(", ", Arrays.toString(validStates.toArray())));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: ditto. please generate the same message as before

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java [hunk 6]
```diff
@@ -213,11 +211,11 @@ void checkArgs() {
         if (options.has(describeOpt)) {
             if (!options.has(groupOpt) && !options.has(allGroupsOpt))
                 CommandLineUtils.printUsageAndExit(parser,
-                    "Option " + describeOpt + " takes one of these options: " + join(allGroupSelectionScopeOpts, ", "));
+                    "Option " + describeOpt + " takes one of these options: " + String.join(", ", Arrays.toString(allGroupSelectionScopeOpts.toArray())));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java]: ditto

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java [hunk 7]
```diff
@@ -213,11 +211,11 @@ void checkArgs() {
...
             List<OptionSpec<?>> mutuallyExclusiveOpts = Arrays.asList(membersOpt, offsetsOpt, stateOpt);
             if (mutuallyExclusiveOpts.stream().mapToInt(o -> options.has(o) ? 1 : 0).sum() > 1) {
                 CommandLineUtils.printUsageAndExit(parser,
-                    "Option " + describeOpt + " takes at most one of these options: " + join(mutuallyExclusiveOpts, ", "));
+                    "Option " + describeOpt + " takes at most one of these options: " + String.join(", ", Arrays.toString(mutuallyExclusiveOpts.toArray())));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java]: ditto

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java [hunk 8]
```diff
@@ -230,7 +228,7 @@ void checkArgs() {
         if (options.has(deleteOpt)) {
             if (!options.has(groupOpt) && !options.has(allGroupsOpt))
                 CommandLineUtils.printUsageAndExit(parser,
-                    "Option " + deleteOpt + " takes one of these options: " + join(allGroupSelectionScopeOpts, ", "));
+                    "Option " + deleteOpt + " takes one of these options: " + String.join(", ", Arrays.toString(allGroupSelectionScopeOpts.toArray())));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java]: ditto

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java [hunk 9]
```diff
@@ -239,7 +237,7 @@ void checkArgs() {
         if (options.has(deleteOffsetsOpt)) {
             if (!options.has(groupOpt) || !options.has(topicOpt))
                 CommandLineUtils.printUsageAndExit(parser,
-                    "Option " + deleteOffsetsOpt + " takes the following options: " + join(allDeleteOffsetsOpts, ", "));
+                    "Option " + deleteOffsetsOpt + " takes the following options: " + String.join(", ", Arrays.toString(allDeleteOffsetsOpts.toArray())));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java]: ditto

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java [hunk 10]
```diff
@@ -255,7 +253,7 @@ void checkArgs() {
 
             if (!options.has(groupOpt) && !options.has(allGroupsOpt))
                 CommandLineUtils.printUsageAndExit(parser,
-                    "Option " + resetOffsetsOpt + " takes one of these options: " + join(allGroupSelectionScopeOpts, ", "));
+                    "Option " + resetOffsetsOpt + " takes one of these options: " + String.join(", ", Arrays.toString(allGroupSelectionScopeOpts.toArray())));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java]: ditto

File: clients/src/main/java/org/apache/kafka/clients/admin/ConsumerGroupDescription.java [hunk 11]
```diff
@@ -161,7 +160,7 @@ public  Set<AclOperation> authorizedOperations() {
     public String toString() {
         return "(groupId=" + groupId +
             ", isSimpleConsumerGroup=" + isSimpleConsumerGroup +
-            ", members=" + Utils.join(members, ",") +
+            ", members=" + String.join(",", Arrays.toString(members.toArray())) +
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/ConsumerGroupDescription.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@chiacyu thanks for this cleanup, but please keep the same message as much as possible.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 2 above)
Comment (chiacyu) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I'm thinking to append `replace("[", "").replace("]", "")` to eliminate brackets. If these's any better representations please let me know. Thanks!

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: how about `partitions.stream().map(TopicPartition::toString).collect(Collectors.joining(","))`?

Comment (chiacyu) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Got it, thanks for the advice!

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 12]
```diff
@@ -584,27 +584,6 @@ public static String formatBytes(long bytes) {
...
-     * @return The string representation.
-     */
-    public static <T> String join(Collection<T> collection, String separator) {
-        Objects.requireNonNull(collection);
-        return mkString(collection.stream(), "", "", separator);
-    }
-
     /**
      * Create a string representation of a stream surrounded by `begin` and `end` and joined by `separator`.
      *
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: It seems `mkString(Stream<T> stream, String begin, String end, String separator)` can be removed too since the only one use case (https://github.com/apache/kafka/blob/5de5d967adffd864bad3ec729760a430253abf38/tools/src/test/java/org/apache/kafka/tools/consumer/group/ConsumerGroupServiceTest.java#L199) does not use "begin" and "end"

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java [hunk 13]
```diff
@@ -131,7 +130,7 @@ public class CommonClientConfigs {
 
     public static final String SECURITY_PROTOCOL_CONFIG = "security.protocol";
     public static final String SECURITY_PROTOCOL_DOC = "Protocol used to communicate with brokers. Valid values are: " +
-        Utils.join(SecurityProtocol.names(), ", ") + ".";
+        String.join(", ", SecurityProtocol.names()).replace("[", "").replace("]", "") + ".";
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java]: please don't use `replace`. We can generate the correct message at once.

File: clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java [hunk 14]
```diff
@@ -392,14 +392,14 @@ private String topicPartitionsToLogString(Collection<TopicPartition> partitions)
         if (!log.isTraceEnabled()) {
             return String.format("%d partition(s)", partitions.size());
         }
-        return "(" + Utils.join(partitions, ", ") + ")";
+        return "(" + String.join(", ", Arrays.toString(partitions.toArray())) + ")";
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java]: ditto

File: clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java [hunk 15]
```diff
@@ -392,14 +392,14 @@ private String topicPartitionsToLogString(Collection<TopicPartition> partitions)
...
     }
 
     private String topicIdPartitionsToLogString(Collection<TopicIdPartition> partitions) {
         if (!log.isTraceEnabled()) {
             return String.format("%d partition(s)", partitions.size());
         }
-        return "(" + Utils.join(partitions, ", ") + ")";
+        return "(" + String.join(", ", Arrays.toString(partitions.toArray())) + ")";
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@chiacyu please fix all similar issues - `Arrays.toString` generate extra `[]` and that can cause different message

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 12 above)
Comment (chiacyu) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: I looked up the code base and found there're almost dozens use cases of `Utils.mkString()`. Should we remove that too? If so, is that a better representation to eliminate it in another PR? Thanks!

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: There are only 11 use cases in code base, so we can do the cleanup in this PR.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 16]
```diff
@@ -1051,13 +1051,13 @@ public static ValidString in(String... validStrings) {
         public void ensureValid(String name, Object o) {
             String s = (String) o;
             if (!validStrings.contains(s)) {
-                throw new ConfigException(name, o, "String must be one of: " + Utils.join(validStrings, ", "));
+                throw new ConfigException(name, o, "String must be one of: " + String.join(",", validStrings));
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: `,` -> `, `

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 17]
```diff
@@ -1051,13 +1051,13 @@ public static ValidString in(String... validStrings) {
...
             }
 
         }
 
         public String toString() {
-            return "[" + Utils.join(validStrings, ", ") + "]";
+            return "[" + String.join(",", validStrings) + "]";
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: ditto

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 18]
```diff
@@ -1079,12 +1079,12 @@ public static CaseInsensitiveValidString in(String... validStrings) {
         public void ensureValid(String name, Object o) {
             String s = (String) o;
             if (s == null || !validStrings.contains(s.toUpperCase(Locale.ROOT))) {
-                throw new ConfigException(name, o, "String must be one of (case insensitive): " + Utils.join(validStrings, ", "));
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:1082]: ditto

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 19]
```diff
@@ -1079,12 +1079,12 @@ public static CaseInsensitiveValidString in(String... validStrings) {
...
+                throw new ConfigException(name, o, "String must be one of (case insensitive): " + String.join(",", validStrings));
             }
         }
 
         public String toString() {
-            return "(case insensitive) [" + Utils.join(validStrings, ", ") + "]";
+            return "(case insensitive) [" + String.join(",", validStrings) + "]";
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: ditto

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 20]
```diff
@@ -1205,7 +1205,8 @@ public void ensureValid(String name, Object value) {
             }
 
             if (!foundIllegalCharacters.isEmpty()) {
-                throw new ConfigException(name, value, "String may not contain control sequences but had the following ASCII chars: " + Utils.join(foundIllegalCharacters, ", "));
+                throw new ConfigException(name, value, "String may not contain control sequences but had the following ASCII chars: " +
+                        foundIllegalCharacters.stream().map(Object::toString).collect(Collectors.joining(",")));
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@chiacyu Thanks for your patch. It would be great to keep the same message as before. Hence, please take a look at following comments.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 12 above)
Comment (chiacyu) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Thanks for the reply. Suggestions applied.

File: clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java [hunk 21]
```diff
@@ -25,21 +25,21 @@
 import org.apache.kafka.common.requests.FetchRequest.PartitionData;
 import org.apache.kafka.common.requests.FetchResponse;
 import org.apache.kafka.common.utils.LogContext;
-import org.apache.kafka.common.utils.Utils;
 import org.slf4j.Logger;
 
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
+import java.util.LinkedHashMap;
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java]: Could you reduce the changes of imports? Also, the new order is weird to me as it is not a lexicographical order

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 22]
```diff
@@ -399,7 +398,7 @@ public TopicCreationResponse createOrFindTopics(NewTopic... topics) {
             }
         }
         if (topicsByName.isEmpty()) return EMPTY_CREATION;
-        String topicNameList = Utils.join(topicsByName.keySet(), "', '");
+        String topicNameList = String.join("', '", topicsByName.keySet()).replace("[", "").replace("]", "");
```
Comment (chia7712) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: Why we need those `replace`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 23]
```diff
@@ -475,7 +474,7 @@ public Map<String, TopicDescription> describeTopics(String... topics) {
         if (topics == null) {
             return Collections.emptyMap();
         }
-        String topicNameList = String.join(", ", topics);
+        String topicNameList = String.join(", ", topics).replace("[", "").replace("]", "");
```
Comment (chia7712) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: ditto

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 24]
```diff
@@ -630,7 +629,7 @@ public Map<String, Config> describeTopicConfigs(String... topicNames) {
         if (topics.isEmpty()) {
             return Collections.emptyMap();
         }
-        String topicNameList = String.join(", ", topics);
+        String topicNameList = String.join(", ", topics).replace("[", "").replace("]", "");
```
Comment (chia7712) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: ditto

File: server/src/main/java/org/apache/kafka/server/config/ReplicationConfigs.java [hunk 25]
```diff
@@ -116,7 +115,7 @@ public class ReplicationConfigs {
     public static final String INTER_BROKER_SECURITY_PROTOCOL_DEFAULT = SecurityProtocol.PLAINTEXT.toString();
     public static final String INTER_BROKER_LISTENER_NAME_CONFIG = "inter.broker.listener.name";
     public static final String INTER_BROKER_SECURITY_PROTOCOL_DOC = "Security protocol used to communicate between brokers. Valid values are: " +
-            Utils.join(SecurityProtocol.names(), ", ") + ". It is an error to set this and " + INTER_BROKER_LISTENER_NAME_CONFIG +
+            String.join(", ", SecurityProtocol.names()).replace("[", "").replace("]", "") + ". It is an error to set this and " + INTER_BROKER_LISTENER_NAME_CONFIG +
```
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/ReplicationConfigs.java]: ditto

File: storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java [hunk 26]
```diff
@@ -216,7 +217,7 @@ public List<ConsumerRecord<String, String>> consume(TopicPartition topicPartitio
         Function0<String> messageSupplier = () ->
                 String.format("Could not consume %d records of %s from offset %d in %d ms. %d message(s) consumed:%s%s",
                         expectedTotalCount, topicPartition, fetchOffset, timeoutMs, records.size(), sep,
-                        Utils.join(records, sep));
+                        String.join(sep, Arrays.toString(records.toArray())));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java]: `records.stream().map(Object::toString).collect(Collectors.joining(","))`

File: tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java [hunk 27]
```diff
@@ -230,8 +231,8 @@ else if (leader.lastCaughtUpTimestamp().isPresent() && maxLagFollower.lastCaught
             "\nHighWatermark:          " + quorumInfo.highWatermark() +
             "\nMaxFollowerLag:         " + maxFollowerLag +
             "\nMaxFollowerLagTimeMs:   " + maxFollowerLagTimeMs +
-            "\nCurrentVoters:          " + Utils.mkString(quorumInfo.voters().stream().map(v -> v.replicaId()), "[", "]", ",") +
-            "\nCurrentObservers:       " + Utils.mkString(quorumInfo.observers().stream().map(v -> v.replicaId()), "[", "]", ",")
+            "\nCurrentVoters:          " + "[" + quorumInfo.voters().stream().map(QuorumInfo.ReplicaState::replicaId).map(Object::toString).collect(Collectors.joining(",")) + "]" +
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java]: `quorumInfo.voters().stream().map(QuorumInfo.ReplicaState::replicaId).map(Object::toString).collect(Collectors.joining(",", "[", "]"))`

File: tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java [hunk 28]
```diff
@@ -230,8 +231,8 @@ else if (leader.lastCaughtUpTimestamp().isPresent() && maxLagFollower.lastCaught
...
+            "\nCurrentObservers:       " + "[" + quorumInfo.observers().stream().map(QuorumInfo.ReplicaState::replicaId).map(Objects::toString).collect(Collectors.joining(",")) + "]"
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java]: ditto

File: tools/src/main/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommand.java [hunk 29]
```diff
@@ -488,12 +488,12 @@ private static void clearAllThrottles(Admin adminClient,
         targetParts.forEach(t -> brokers.addAll(t.getValue()));
 
         System.out.printf("Clearing broker-level throttles on broker%s %s%n",
-            brokers.size() == 1 ? "" : "s", Utils.join(brokers, ","));
+            brokers.size() == 1 ? "" : "s", String.join(",", Arrays.toString(brokers.toArray())));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommand.java]: `brokers.stream().map(Object::toString).collect(Collectors.joining(","))`

File: trogdor/src/main/java/org/apache/kafka/trogdor/fault/ProcessStopFaultWorker.java [hunk 30]
```diff
@@ -80,7 +80,7 @@ private void sendSignals(Platform platform, String signalName) throws Exception
                 id, javaProcessName, signalName);
         } else {
             log.info("{}: sending {} to {} pid(s) {}",
-                id, signalName, javaProcessName, Utils.join(pids, ", "));
+                id, signalName, javaProcessName, String.join(", ", Arrays.toString(pids.toArray())));
```
Comment (chia7712) [inline trogdor/src/main/java/org/apache/kafka/trogdor/fault/ProcessStopFaultWorker.java]: `pids.stream().map(Object::toString).collect(Collectors.joining(","))`

File: trogdor/src/main/java/org/apache/kafka/trogdor/workload/RoundTripWorker.java [hunk 31]
```diff
@@ -318,7 +320,7 @@ void log() {
                 }
             }
             log.info("{}: consumer waiting for {} message(s), starting with: {}",
-                id, numToReceive, Utils.join(list, ", "));
+                id, numToReceive, String.join(", ", Arrays.toString(list.toArray())));
```
Comment (chia7712) [inline trogdor/src/main/java/org/apache/kafka/trogdor/workload/RoundTripWorker.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@chiacyu thanks for updated PR. a couple of comments left. PTAL

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java (hunk 22 above)
Comment (chiacyu) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: Sorry just missed that, would fix soon. Thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java [hunk 32]
```diff
@@ -466,7 +467,7 @@ private static ConfigDef config(Crypto crypto) {
                                         + "compatibility");
                             }
                         },
-                        () -> "[" + Utils.join(ConnectProtocolCompatibility.values(), ", ") + "]"),
+                        () -> "[" + String.join(", ", Arrays.toString(ConnectProtocolCompatibility.values())) + "]"),
```
Comment (chia7712) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java]:
```java
                        () -> Arrays.stream(ConnectProtocolCompatibility.values()).map(ConnectProtocolCompatibility::toString)
                                .collect(Collectors.joining(", ", "[", "]"))),
```

**Review by chia7712 — COMMENTED** [review]
@chiacyu thanks for your patience. one last comment. PTAL

File: tools/src/main/java/org/apache/kafka/tools/VerifiableConsumer.java [hunk 33]
```diff
@@ -536,7 +537,7 @@ private static ArgumentParser argParser() {
                 .setDefault(ConsumerConfig.DEFAULT_GROUP_PROTOCOL)
                 .metavar("GROUP_PROTOCOL")
                 .dest("groupProtocol")
-                .help(String.format("Group protocol (must be one of %s)", Utils.join(GroupProtocol.values(), ", ")));
+                .help(String.format("Group protocol (must be one of %s)", String.join(", ", Arrays.toString(GroupProtocol.values()))));
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/VerifiableConsumer.java]: `Arrays.stream(GroupProtocol.values()).map(Object::toString).collect(Collectors.joining(", "))`

**Review by chia7712 — COMMENTED** [review]
@chiacyu one more comment. PTAL

**Review by chia7712 — APPROVED** [review]
LGTM, will merge it if no related test error

**Comment by chia7712** [comment]
```
./gradlew cleanTest :tools:test --tests MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful :storage:test --tests TransactionsWithTieredStoreTest.testFailureToFenceEpoch :connect:runtime:test --tests org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testResetSinkConnectorOffsetsOverriddenConsumerGroupId :metadata:test --tests QuorumControllerTest.testConfigResourceExistenceChecker --tests QuorumControllerTest.testFenceMultipleBrokers :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :connect:mirror:test --tests MirrorConnectorsIntegrationBaseTest.testReplicateSourceDefault :core:test --tests ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit

```
no related error, and they pass on my local.

**Comment by marcelstoer** [comment]
IMO it's really bad practice to alter public methods in a minor release. Why would you not want to follow SemVer?

**Comment by chia7712** [comment]
> IMO it's really bad practice to alter public methods in a minor release. Why would you not want to follow SemVer?

Pardon me, could you clarify the details about the broken public APIs caused by this PR? Just to note, `Utils.java` is not part of the public APIs since it resides in the non-public package `org.apache.kafka.common.utils`

**Comment by chia7712** [comment]
As referenced in [this line](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/package-info.java#L19)

> This package is not a supported Kafka API; the implementation may change without warning between minor or patch releases.

**Comment by marcelstoer** [comment]
Thanks for your feedback.

When you stated in your first comment that the `utils` package be non-public I went looking for any sort of documented confirmation. That's when I found the `package-info.java`.

I guess that's unfortunate naming then. It isn't exactly [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) to find a sub package in a `common` package to be internal (and not be called `internal`). But well, it is what it is. I'll try to remember that.
